lib_eio/sync.ml

```diff
@@ -0,0 +1,448 @@
+  | `Cancelled _ -> failwith "Already cancelled!"
+  | `Resuming -> false  (* Cancellation fails for now, but we remember we wanted to cancel. *)
+  | `Suspended cancel -> cancel ()
+
+(* Putting. *)
+
+(* Like [add_to_cell], but we haven't created our value yet as we haven't suspended the fiber. *)
+let rec producer_resume_cell t ~success ~in_transition cell =
+  match Atomic.get (cell : _ Cell.t Atomic.t) with
+  | Item _ -> assert false
+  | In_transition -> in_transition cell
+  | Finished -> producer_resume_cell t ~success ~in_transition (Q.next_resume t.consumers)
+  | Slot k as old ->
+    if Atomic.compare_and_set cell old Finished then success k
+    else producer_resume_cell t ~success ~in_transition cell
+
+(* This is essentially the main [put] function, but parameterised so it can be shared with
+   the rejoin-after-rejection case. *)
+let producer_join (t : _ t) ~success ~suspend =
+  let old = Atomic.fetch_and_add t.balance (+1) in
+  if old < 0 then (
+    let cell = Q.next_resume t.consumers in
+    producer_resume_cell t cell
+      ~success
+      ~in_transition:(fun cell -> suspend (Short cell))
+  ) else (
+    suspend (Long (Q.next_suspend t.producers))
+  )
+
+(* Called when a consumer took our value but then rejected it.
+   We start the put operation again, except that our fiber is already suspended
+   so no need to do that again. We're probably running in the consumer's domain
+   (unless the consumer provided their callback while we were cancelling). *)
+let put_already_suspended t request =
+  producer_join t
+    ~success:(exchange request)
+    ~suspend:(fun loc ->
+        let Short cell | Long (_, cell) = loc in
+        add_to_cell t.consumers (Item request) cell;
+        let rec aux () =
+          match Atomic.get request.cancel, loc with
+          | (`Suspended _ | `Resuming as prev), Long loc ->
+            (* We might be suspended for a while. Update the cancel function with the new location. *)
+            let cancel_fn () = cancel t loc in
+            if not (Atomic.compare_and_set request.cancel prev (`Suspended cancel_fn)) then aux ()
+          | `Cancelled ex, Long loc ->
+            (* We got cancelled after the peer removed our cell and before we updated the
+               cancel function with the new location, or we were cancelled while doing a
+               (non-cancellable) resume. Deal with it now. *)
+            if cancel t loc then request.kp (Error ex);
+            (* else we got resumed first *)
+          | _, Short _ ->
+            (* We can't cancel while in the process of resuming a cell on the [consumers] queue.
+               We could set [cancel] to [`Resuming] here, but there's no need as trying to use the
+               old cancel function will find the old cell is cancelled and set [request.cancel]
+               to [`Cancelled]), as required. *)
+            ()
+        in aux ()
+      )
+
+(* We tried to [put] and no value was immediately available.
```

Comment:
I guess this should say "no ~value~slot was immediately available".

Reply:
Yes, that would be clearer. Note that I am using "value" to mean "item or slot" in some places (e.g. in `add_to_cell`).